File: 1777-determine-if-two-strings-are-close/solution.java
class Solution {
    public boolean closeStrings(String word1, String word2) {
        // 1. בדיקת אורך בסיסית
        if (word1.length() != word2.length()) {
            return false;
        }

        // יצירת שתי מפות לתדירויות
        Map<Character, Integer> map1 = new HashMap<>();
        Map<Character, Integer> map2 = new HashMap<>();

        for (char c : word1.toCharArray()) {
            map1.put(c, map1.getOrDefault(c, 0) + 1);
        }
        for (char c : word2.toCharArray()) {
            map2.put(c, map2.getOrDefault(c, 0) + 1);
        }

        // 2. בדיקה ששתי המחרוזות מכילות את אותו סט של אותיות ייחודיות
        // (keySet() מחזיר Set של כל המפתחות במפה)
        if (!map1.keySet().equals(map2.keySet())) {
            return false;
        }

        // 3. בדיקה שהתפלגות התדירויות זהה
        // אנחנו מוציאים את הערכים (הכמויות), ממיינים אותם ומשווים
        List<Integer> values1 = new ArrayList<>(map1.values());
        List<Integer> values2 = new ArrayList<>(map2.values());

        Collections.sort(values1);
        Collections.sort(values2);

        return values1.equals(values2);
    }
}

File: 1777-determine-if-two-strings-are-close/solution.py
class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        if len(word1) != len(word2):
            return False

        freq1 = {}
        for char in word1:
            freq1[char] = freq1.get(char, 0) + 1

        freq2 = {}
        for char in word2:
            freq2[char] = freq2.get(char, 0) + 1

        if set(freq1.keys()) != set(freq2.keys()):
            return False

        vals1 = sorted(freq1.values())
        vals2 = sorted(freq2.values())

        return vals1 == vals2
